storey/sources.py

```diff
@@ -619,56 +617,87 @@ def __init__(
             kwargs["id_field"] = id_field
         _IterableSource.__init__(self, **kwargs)
         WithUUID.__init__(self)
+        #  in order to raise exception also for key_field=0
+        if key_field is not None:
+            key_fields = [key_field] if not isinstance(key_field, list) else key_field
+            if any([not isinstance(single_key_field, str) for single_key_field in key_fields]):
+                raise ValueError("key_field should be string/list of strings only!")
+        if id_field and not isinstance(id_field, str):
+            raise ValueError("id_field should be in string type only!")
+        self._key_field = key_field
+        self._id_field = id_field
         if isinstance(dfs, pandas.DataFrame):
             dfs = [dfs]
+        if dfs:
+            for df in dfs:
+                self._validate_fields(df=df)
         self._dfs = dfs
-        self._key_field = key_field
-        self._id_field = id_field
+
+    def _get_key(self, body: dict):
+        """
+        return key values by columns + column name that have a non value if exists.
+        """
+        key = None
+        if self._key_field:
+            if isinstance(self._key_field, list):
+                key = []
+                for key_field in self._key_field:
+                    single_key = body[key_field]
+                    if pd.isna(single_key):
+                        return single_key, key_field
+                    key.append(single_key)
+            else:
+                key = body[self._key_field]
+                if pd.isna(key):
+                    return key, self._key_field
+        return key, None
 
     async def _run_loop(self):
         for df in self._dfs:
-            for namedtuple in df.itertuples():
-                create_event = True
-                body = namedtuple._asdict()
-                index = body.pop("Index")
-                if len(df.index.names) > 1:
-                    for i, index_column in enumerate(df.index.names):
-                        body[index_column] = index[i]
-                elif df.index.names[0] is not None:
-                    body[df.index.names[0]] = index
-
-                key = None
-                if self._key_field:
-                    if isinstance(self._key_field, list):
-                        key = []
-                        for key_field in self._key_field:
-                            if key_field not in body or pandas.isna(body[key_field]):
-                                create_event = False
-                                break
-                            key.append(body[key_field])
-                    else:
-                        key = body[self._key_field]
-                        if key is None:
-                            create_event = False
-                if create_event:
+            columns = list(df.columns)
+            if not df.index.empty and not (len(df.index.names) == 1 and df.index.names[0] is None):
```

Comment:
Probably clearer interntion:
```suggestion
            is_df_index_nonempty = not df.index.empty and not (len(df.index.names) == 1 and df.index.names[0] is None)
            if is_df_index_nonempty:
```